lib/spack/spack/package.py

```diff
@@ -1245,43 +1245,68 @@ def do_fake_install(self):
         dump_packages(self.spec, packages_dir)
 
     def _if_make_target_execute(self, target):
-        try:
-            # Check if we have a makefile
-            file = [x for x in ('Makefile', 'makefile') if os.path.exists(x)]
-            file = file.pop()
-        except IndexError:
+        make = inspect.getmodule(self).make
+
+        # Check if we have a Makefile
+        for makefile in ['GNUmakefile', 'Makefile', 'makefile']:
+            if os.path.exists(makefile):
+                break
+        else:
             tty.msg('No Makefile found in the build directory')
             return
 
-        # Check if 'target' is in the makefile
-        regex = re.compile('^' + target + ':')
-        with open(file, 'r') as f:
-            matches = [line for line in f.readlines() if regex.match(line)]
-
-        if not matches:
-            tty.msg("Target '" + target + ":' not found in Makefile")
+        # Check if 'target' is a valid target
+        #
+        # -q, --question
+        #       ``Question mode''. Do not run any commands, or print anything;
+        #       just return an exit status that is zero if the specified
+        #       targets are already up to date, nonzero otherwise.
+        #
+        # https://www.gnu.org/software/make/manual/html_node/Options-Summary.html
+        #
+        # The exit status of make is always one of three values:
+        #
+        # 0     The exit status is zero if make is successful.
+        #
+        # 2     The exit status is two if make encounters any errors.
+        #       It will print messages describing the particular errors.
+        #
+        # 1     The exit status is one if you use the '-q' flag and make
+        #       determines that some target is not already up to date.
+        #
+        # https://www.gnu.org/software/make/manual/html_node/Running.html
+        #
+        # NOTE: This only works for GNU Make, not NetBSD Make.
```

Comment:
When you say:

> The downside to this approach is that if someone runs spack install --test on a system that uses BSD Make, it will always think that a check target exists and run it even if it doesn't.

How does this compare to the previous behavior for BSD make in Spack?

What is the exit status if you run this for BSD make? Can you distinguish a failing BSD make from a GNU make that is failing because the target doesn't exist?

Reply:
Previously, we would grep the file for a `^check:` target. It didn't matter if it was BSD or GNU Make, it would behave identically. The problem was that this regex isn't powerful enough to capture all possible `check` targets (see first paragraph of PR description).

The current behavior is that it is guaranteed to work perfectly for GNU Make but doesn't work at all for BSD Make.

With BSD Make, there is no separate exit status for "this target doesn't exist", it's the same exit status as "the target has not yet been built", so there is no way to distinguish by exit status alone. We could fall back on a regex if we need to support BSD Make, although I've never encountered a system that uses BSD Make.